Redis implementation

mdiin opened the pull request:
This PR implements the atomic, value, and volatile reference types with a Redis backed. Carmine was chosen for the fact that it is the simplest to work with given the constraints, i.e. atomic operations, and because nippy was already a dependency of this library.

I would like to discuss a merge of this into master, as well as a design decision I made: That keys can be namespaced. On the surface this seems like a necessary decision if we want to store large maps that receive frequent updates, however I do not know if in reality it is a premature optimization.

Tests are still missing.

Commit:
Add carmine dependency for Redis backend

Commit:
Add Redis-backed implementation for atomic, value and volatile.

Commit:
Update ns doc string.

wotbrew commented:
@mdiin Thanks for this, I'll review asap.

wotbrew commented on src/riverford/durable_ref/scheme/redis/carmine.clj:
```diff
@@ -0,0 +1,152 @@
+
+(defn- get-credentials
+  [opts connection]
+  (let [c1 (:credentials opts)
```
I think I would prefer `(-> opts :scheme :redis :carmine :credentials)` to be consistent with the Dynamo/S3 scheme option passing style.

mdiin replied:
Of course, I actually intended to follow that style, but that must have slipped.

wotbrew commented on src/riverford/durable_ref/scheme/redis/carmine.clj:
```diff
@@ -0,0 +1,152 @@
+        conn (-> connection
+                 (assoc-in [:spec :password] (or (:password credentials) ""))
+                 (assoc-in [:spec :db] db))]
+    (let [[_ [_ res-raw]] (car/atomic conn 100
```
I think this will throw after 100 failed attempts under a contention scenario. I would prefer the default to be to keep trying in the event of CAS/optimistic lock failures, and to only throw on an unrelated exception (such as a connection issue). Clojure's atom will spin forever until it completes the swap, and I would like to carry that expectation over.  However, I do see the value in being able to supply an attempt limit as an opt - that would be cool.

mdiin replied:
The problem here, I think, is that there can be a lot more pressure on a distributed atom than on a local one. Regardless whether it is a good idea or not to let it spin forever, this will need some more thinking and research on the Redis side of things; I have to figure out how to go about implementing the same functionality without the Carmine `atomic` macro.

I will start looking into this, but it will take some more time. :-)

wotbrew replied:
I think being able to provide spin policy (back-off config, limits, etc) via opts is a good idea - as long as the default is to spin forever so that IAtom meets expectations. Regardless of spin timeouts you want to be sure you are not using this lib in high contention scenarios as optimistic control like this never works well there. As for Redis I don't know I'm afraid would need to look into myself, I'll wait to see what you can come up with!

wotbrew reviewed (changes requested):
Looks really cool, good job - a couple of suggested tweaks and then happy to merge in. Tests would be welcome too, should be easy enough to copy some tests such as the dynamo integration tests. Just run them if the required env vars are present, e.g DURABLE_REF_TEST_REDIS_HOST.

mdiin commented:
Thanks for looking into this so quickly! I will address your review comments soon.

Commit:
Change credentials scheme as requested by @danstone

mdiin commented:
I think I will remove the key nesting functionality, as it differs from the other reference types, and does not really fulfill the need I had thought it would.

Commit:
Drop key-namespacing from Carmine implementation.

mdiin commented:
@danstone As you might have guessed, work has had me side-tracked. I probably won't get the time to return to this PR any time soon, if at all, so I will close it for now.

Anyway, I get the feeling that my use-case will involve too many writes for a Redis implementation to be viable. If/when work takes me back to the code where I considered this approach, I will reevaluate.

mdiin committed:
Allow specifying `nil` for password

Not all redis instances are password protected, e.g. development
instances or those just created to test durable-ref.

mdiin committed:
Default `do-atomic-swap!` to spin forever

As suggested in the Carmine sources I opted for a custom CAS operation
implemented using the Redis primitives.

mdiin commented:
Work still doesn't have me back on this, but my interest is there. I have mostly completed the swap implementation now, so I'm reopening this. It's still a work in progress though.

mdiin committed:
Add tests for Redis/carmine scheme

These are simple copies of the amazonica integration tests for s3 and
dynamodb.

mdiin committed:
Readme update for Redis scheme

wotbrew commented:
Looking good, I'll wait until you are ready and then review, looking forward to merging this!

mdiin committed:
Use `:cas-back-off-fn` in `dref/do-atomic-swap!`

This is a more general solution than simply specifying a maximum number
of retries, and mirrors the solution in the DynamoDB scheme.

mdiin committed:
Create Carmine connection in `let`

This is just for consistency and ease of reading. The arrow chaining
removes the noise in the `conn` specifications.

mdiin commented:
@danstone I think the PR is ready for review now. Let me know what you think.

wotbrew commented:
Thanks for this, just had a quick scan and looks good, if I could have a bit more time this week just to digest it and test it then I'll get back to you / merge.

wotbrew commented:
@mdiin Merged, I'll do a release to clojars a bit later today, I would normally do an announcement in slack and on the mailing list if you are happy to be attributed?

Thanks again for this, really great work!

mdiin commented:
@danstone Cool, I thanks! Feel free to make the announcement and mention me whenever you feel like it. :-)
